File: Kidney John/UIActivity/KJVideoFavouriteActivity.m
//
//  KJVideoFavouriteActivity.m
//  Kidney John
//
//  Created by jl on 5/05/2014.
//  Copyright (c) 2014 Josh Lapham. All rights reserved.
//

#import "KJVideoFavouriteActivity.h"
#import "KJVideoStore.h"
#import "KJVideo.h"
#import "NSUserDefaults+KJSettings.h"
#import "KJParseAnalyticsStore.h"

@implementation KJVideoFavouriteActivity {
    NSString *titleOfActivity;
    KJVideo *videoObject;
}

#pragma mark - Init method

- (id)initWithVideo:(KJVideo *)video {
    self = [super init];
    if (self) {
        // Init video object
        videoObject = video;
        
        // Init activity title, depending on favourite status of video
        if (!videoObject.isFavourite) {
            titleOfActivity = NSLocalizedString(@"Add To Favourites", @"Title of button to favourite an item");
        }
        else {
            titleOfActivity = NSLocalizedString(@"Remove From Favourites", @"Title of button to remove an item as a favourite");
        }
    }
    
    return self;
}

#pragma mark - Other methods

- (NSString *)activityType {
    return @"com.joshlapham.Kidney-John favourite video";
}

- (NSString *)activityTitle {
    return titleOfActivity;
}

- (UIImage *)activityImage {
    if (!videoObject.isFavourite) {
        return [UIImage imageNamed:@"add-to-fav.png"];
    }
    else {
        return [UIImage imageNamed:@"remove-from-fav.png"];
    }
}

- (BOOL)canPerformWithActivityItems:(NSArray *)activityItems {
    return YES;
}

- (void)prepareWithActivityItems:(NSArray *)activityItems {
    // Toggle favourite status for videoObject
    videoObject.isFavourite = !videoObject.isFavourite;
    
    // Track action with Parse analytics (if enabled)
    if ([NSUserDefaults kj_shouldTrackFavouritedItemEventsWithParseSetting]) {
        [[KJParseAnalyticsStore sharedStore] trackVideoFavouriteEventForVideo:videoObject];
    }
    
    // Save managedObjectContext
    NSError *error;
    if (![videoObject.managedObjectContext save:&error]) {
        DDLogError(@"%s - failed to save managedObjectContext: %@", __func__, [error debugDescription]);
    }
    
    else {
        DDLogInfo(@"%s - saved managedObjectContext", __func__);
    }
}

- (UIViewController *)activityViewController {
    return nil;
}

- (void)performActivity {
    [self activityDidFinish:YES];
}

@end

File: Kidney John/UIActivity/KJVideoFavouriteActivity.swift
//
//  KJVideoFavouriteActivity.swift
//  Kidney John
//
//  Created by Josh Lapham on 15/09/2015.
//  Copyright © 2015 Josh Lapham. All rights reserved.
//

import UIKit
import CocoaLumberjackSwift

class KJVideoFavouriteActivity: UIActivity {
    var titleOfActivity: String
    var videoObject: KJVideo
    
    // MARK: UIActivity
    override func activityType() -> String? {
        return "com.joshlapham.Kidney-John favourite video"
    }
    
    override func activityTitle() -> String? {
        return self.titleOfActivity
    }
    
    override func activityImage() -> UIImage? {
        if self.videoObject.isFavourite == false {
            return UIImage(named: "add-to-fav.png")
        } else {
            return UIImage(named: "remove-from-fav.png")
        }
    }
    
    override func canPerformWithActivityItems(activityItems: [AnyObject]) -> Bool {
        return true
    }
    
    override func activityViewController() -> UIViewController? {
        return nil
    }
    
    override func performActivity() {
        self.activityDidFinish(true)
    }
    
    override func prepareWithActivityItems(activityItems: [AnyObject]) {
        // Toggle favourite status for videoObject
        videoObject.isFavourite = !videoObject.isFavourite;
        
        // TODO: add/remove item to/from Spotlight API index
        
        // Track action with Parse analytics (if enabled)
        // TODO: revise this after CloudKit refactor
        //        if NSUserDefaults.kj_shouldTrackFavouritedItemEventsWithParseSetting() == true {
        //            KJParseAnalyticsStore.sharedStore().trackVideoFavouriteEventForVideo(self.videoObject)
        //        }
        
        do {
            try self.videoObject.managedObjectContext?.save()
            DDLogVerbose("Saved managedObjectContext")
            
        } catch let error as NSError {
            DDLogError("Failed to save managedObjectContext : \(error.debugDescription)")
        }
    }
    
    // MARK: NSObject
    required init(video: KJVideo) {
        self.videoObject = video
        
        if self.videoObject.isFavourite == false {
            self.titleOfActivity = NSLocalizedString("Add To Favourites", comment: "Title of button to favourite an item")
        } else {
            self.titleOfActivity = NSLocalizedString("Remove From Favourites", comment: "Title of button to remove an item as a favourite")
        }
        
        super.init()
    }
}
